airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java

```diff
@@ -0,0 +1,179 @@
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.Field;
+import com.google.cloud.bigquery.Field.Builder;
+import com.google.cloud.bigquery.Field.Mode;
+import com.google.cloud.bigquery.FieldList;
+import com.google.cloud.bigquery.Schema;
+import com.google.cloud.bigquery.StandardSQLTypeName;
+import com.google.common.base.Preconditions;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.commons.util.MoreIterators;
+import io.airbyte.integrations.base.AirbyteMessageConsumer;
+import io.airbyte.integrations.base.AirbyteStreamNameNamespacePair;
+import io.airbyte.integrations.base.Destination;
+import io.airbyte.integrations.base.IntegrationRunner;
+import io.airbyte.integrations.base.JavaBaseConstants;
+import io.airbyte.protocol.models.AirbyteMessage;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BigQueryDenormalizedDestination extends BigQueryDestination {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(BigQueryDenormalizedDestination.class);
+
+  protected static final String PROPERTIES_FIELD = "properties";
+  protected static final String NESTED_ARRAY_FIELD = "value";
+  private static final String TYPE_FIELD = "type";
+
+  @Override
+  protected String getTargetTableName(String streamName) {
+    // This BigQuery destination does not write to a staging "raw" table but directly to a normalized
+    // table
+    return getNamingResolver().getIdentifier(streamName);
+  }
+
+  @Override
+  protected AirbyteMessageConsumer getRecordConsumer(BigQuery bigquery,
+                                                     Map<AirbyteStreamNameNamespacePair, BigQueryWriteConfig> writeConfigs,
+                                                     ConfiguredAirbyteCatalog catalog,
+                                                     Consumer<AirbyteMessage> outputRecordCollector) {
+    return new BigQueryDenormalizedRecordConsumer(bigquery, writeConfigs, catalog, outputRecordCollector, getNamingResolver());
+  }
+
+  @Override
+  protected Schema getBigQuerySchema(JsonNode jsonSchema) {
+    final List<Field> fieldList = getSchemaFields(getNamingResolver(), jsonSchema);
+    if (fieldList.stream().noneMatch(f -> f.getName().equals(JavaBaseConstants.COLUMN_NAME_AB_ID))) {
+      fieldList.add(Field.of(JavaBaseConstants.COLUMN_NAME_AB_ID, StandardSQLTypeName.STRING));
+    }
+    if (fieldList.stream().noneMatch(f -> f.getName().equals(JavaBaseConstants.COLUMN_NAME_EMITTED_AT))) {
+      fieldList.add(Field.of(JavaBaseConstants.COLUMN_NAME_EMITTED_AT, StandardSQLTypeName.TIMESTAMP));
+    }
```

Comment:
Are the above two `if` checks always false? It seems that the original Json schema will never have the two Airbyte columns.

Reply:
the original JSON Schema can have the two airbyte columns if the streams were produced by airbyte and re-used as source streams

Reply:
Oh, I see. Good to know this.